### ISSUE TITLE
color-contrast variable impact

### ISSUE THREAD

**Issue by joshbruning** [dequelabs/axe-core#3069]
Product: axe-core

Expectation: color-contrast impact should reflect actual user impact

Actual: color-contrast impact is always "serious"

Motivation: we wish to use the impact field to prioritize issues with actual user impact over those with little to no actual user impact

axe-core version: 4.2.2

The current (default) success criterion:
Ensure color contrast of at least 4.5:1 for small text or 3:1 for large text, even if text is part of an image. Large text has been defined in the requirements as 18pt (24 CSS pixels) or 14pt bold (19 CSS pixels). Note: Elements found to have a 1:1 ratio are considered "incomplete" and require a manual review.

For example:
- the difference between 2.999:1 and 3:1 may be imperceptible to all users or fall short due to a floating point rounding error
- 2.9:1 may have higher contrast on some displays than 3.1:1 has on others, reducing the set of impacted users
- 24 CSS pixel text with a 3:1 ratio may be less accessible (which passes) than 23 CSS pixel text with a 4.4:1 ratio (which fails).
- 17pt text on some devices is larger than 18pt text on others

As color-contrast is a quite common type of violation, it would be useful to distinguish between issues which should be fixed immediately for high user impact and those which should be fixed to conform to spec. I would expect some errors to be presented as "minor" violations of a users specified settings. It is possible that users be allowed to configure multiple thresholds - for example, I could configure to see "serious" violations at 2.8:1 and worse, "moderate" violations at 2.8-3.0:1, and "minor" at 3.2:1 and worse - allowing me to prioritize a large backlog better - and also catch cases which may be less accessible on many devices, but conform to spec. The idea here is meant to be illustrative - I would welcome anything the team comes up w/ to help with doing this in a single axe run - color-contrast is an expensive check, so running in multiple times with several different configurations (text size, ratios) slows down dev workflow/CI.

**Comment by straker** [comment]
Thanks for this detailed issue. We as of a few releases ago actually support configuring color contrast with different ratios and assigning them different values (related ticket https://github.com/dequelabs/axe-core/issues/2175). So if you wanted to say, give 3 different priorities to color-contrast, you could create 3 custom rules to do that. 

Below is an example of a single custom rule/check that would be assigning contrast ratios that fail 4.5:1 ratio but are above a 4.0:1 to a minor severity since ones that have lower contrast ratios would be more important to fix first.

```js
axe.configure({
  rules: [{
    id: 'color-contrast-minor',
    matches: 'color-contrast-matches',
    any: ['color-contrast-minor']
  }],
  checks: [{
    id: 'color-contrast-minor',
    evaluate: 'color-contrast-evaluate',
    metadata: {
      impact: 'minor'
    },
    options: {
      contrastRatio: {
        normal: {
          expected: 4.5,
          minThreshold: 4.0
        },
        large: {
          expected: 3,
          minThreshold: 2.5
        }
      },
      ignoreUnicode: true,
      ignoreLength: false,
      ignorePseudo: false,
      boldValue: 700,
      boldTextPt: 14,
      largeTextPt: 18,
      shadowOutlineEmMax: 0.1
    }
  }]
});
```

**Comment by joshbruning** [comment]
Wow, thanks! I missed that.

**Comment by straker** [comment]
Forgot to add, the options you can use are listed in the [check metadata](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/color/color-contrast.json#L4-L20) but we also allow `minThreshold` and `maxThreshold` on the contrastRatio.normal/large object.

**Comment by joshbruning** [comment]
Oh sorry, I forgot to ask as well - the color-contrast check takes a very long time for some pages - does this result in the color contrast value being recomputed for all of the elements?

**Comment by straker** [comment]
Unfortunately yes, this would cause the color contrast to run multiple times per node. The way our architecture works we can't run multiple rules / checks at the same time on a single node.

**Comment by straker** [comment]
I'm going to close this issue if that's alright with you. Feel free to reopen (or open a new ticket) if something else comes up.

**Comment by joshbruning** [comment]
Thanks - for posterity sake, I plan to write an axe results JSON postprocessor to patch impact.
